rust/src/delta_datafusion.rs

```diff
@@ -988,6 +999,345 @@ impl TableProviderFactory for DeltaTableFactory {
+                PATH_COLUMN
+            )))?;
+        for path in array {
+            let path = path.ok_or(DeltaTableError::Generic(format!(
+                "{} cannot be null",
+                PATH_COLUMN
+            )))?;
+
+            match actions.remove(path) {
+                Some(action) => files.push(action),
+                None => {
+                    return Err(DeltaTableError::Generic(
+                        "Unable to map __delta_rs_path to action.".to_owned(),
+                    ))
+                }
+            }
+        }
+    }
+    Ok(files)
+}
+
+/// Determine which files contain a record that statisfies the predicate
+pub(crate) async fn find_files_scan<'a>(
+    snapshot: &DeltaTableState,
+    store: ObjectStoreRef,
+    schema: Arc<ArrowSchema>,
+    file_schema: Arc<ArrowSchema>,
+    candidates: Vec<&'a Add>,
+    state: &SessionState,
+    expression: &Expr,
+) -> DeltaResult<Vec<Add>> {
+    let mut candidate_map: HashMap<String, Add> = HashMap::new();
+
+    let table_partition_cols = snapshot
+        .current_metadata()
+        .ok_or(DeltaTableError::NoMetadata)?
+        .partition_columns
+        .clone();
+
+    let mut file_groups: HashMap<Vec<ScalarValue>, Vec<PartitionedFile>> = HashMap::new();
+    for action in candidates {
+        let mut part = partitioned_file_from_action(action, &schema);
+        part.partition_values
+            .push(ScalarValue::Utf8(Some(action.path.clone())));
+
+        file_groups
+            .entry(part.partition_values.clone())
+            .or_default()
+            .push(part);
+
+        candidate_map.insert(action.path.to_owned(), action.to_owned());
+    }
+
+    let mut table_partition_cols = table_partition_cols
+        .iter()
+        .map(|c| Ok((c.to_owned(), schema.field_with_name(c)?.data_type().clone())))
+        .collect::<Result<Vec<_>, ArrowError>>()?;
+    // Append a column called __delta_rs_path to track the file path
+    table_partition_cols.push((PATH_COLUMN.to_owned(), DataType::Utf8));
+
+    let input_schema = snapshot.input_schema()?;
+    let input_dfschema: DFSchema = input_schema.clone().as_ref().clone().try_into()?;
+
+    let predicate_expr = create_physical_expr(
+        &Expr::IsTrue(Box::new(expression.clone())),
+        &input_dfschema,
+        &input_schema,
+        state.execution_props(),
+    )?;
+
+    let parquet_scan = ParquetFormat::new()
+        .create_physical_plan(
+            state,
+            FileScanConfig {
+                object_store_url: store.object_store_url(),
+                file_schema,
+                file_groups: file_groups.into_values().collect(),
+                statistics: snapshot.datafusion_table_statistics(),
+                projection: None,
+                limit: None,
+                table_partition_cols,
+                infinite_source: false,
+                output_ordering: None,
+            },
+            None,
+        )
+        .await?;
```

Comment:
I found this Parquet scan was pulling all the columns, even ones that weren't involved in the predicate. These changes should make it only pull the ones needed for the predicate, plus the path column.

```suggestion
    // Identify which columns we need to project
    let mut used_columns = expression
        .to_columns()
        ?
        .into_iter()
        .map(|column| input_schema.index_of(&column.name))
        .collect::<Result<Vec<usize>, ArrowError>>()
        .unwrap();
    // Add path column
    used_columns.push(input_schema.fields().len());

    let parquet_scan = ParquetFormat::new()
        .create_physical_plan(
            state,
            FileScanConfig {
                object_store_url: store.object_store_url(),
                file_schema,
                file_groups: file_groups.into_values().collect(),
                statistics: snapshot.datafusion_table_statistics(),
                projection: Some(used_columns),
                limit: None,
                table_partition_cols,
                infinite_source: false,
                output_ordering: None,
            },
            None,
        )
        .await?;

    let input_schema = parquet_scan.schema();
    let input_dfschema: DFSchema = input_schema.clone().as_ref().clone().try_into()?;
    let predicate_expr = create_physical_expr(
        &Expr::IsTrue(Box::new(expression.clone())),
        &input_dfschema,
        &input_schema,
        state.execution_props(),
    )?;
```